compiler/vm/vm.nim

```diff
@@ -2362,22 +2353,98 @@ proc rawExecute(c: var TCtx, start: int, sframe: StackFrameIndex): TFullReg =
 
     inc pc
 
-proc execute(c: var TCtx, start: int, frame: sink TStackFrame): PNode {.inline.} =
+type ExecutionResult = object
+  case success: bool
+  of false:
+    stackTrace: SemReport ## The report storing the stack-trace
+    report: SemReport     ## The report detailing the error
+  of true:
+    result: PNode ## The evaluated value/tree (nkEmpty if nothing was returned)
+
+proc execute(c: var TCtx, start: int, frame: sink TStackFrame): ExecutionResult {.inline.} =
   assert c.sframes.len == 0
   c.sframes.add frame
-  result = rawExecute(c, start, 0).regToNode
-  assert c.sframes.len == 1
-  c.sframes.setLen(0)
 
+  var pc = start
+  var sframe = c.sframes.high
+  try:
+    let r = rawExecute(c, pc, sframe).regToNode
+    assert c.sframes.len == 1
+    result = ExecutionResult(success: true, result: r)
+  except VmError as e:
+    # Execution failed, generate a stack-trace and wrap the error payload
+    # into an `ExecutionResult`
+    let trace = createStackTrace(c, sframe, pc)
+    result = ExecutionResult(
+      success: false,
+      stackTrace: trace,
+      report: move e.report)
+
+    if result.report.location.isNone():
+      # Use the location info of the failing instruction if none is provided
+      result.report.location = some(c.debug[pc])
+
+    # Set stack-trace report information
+    result.stackTrace.location = some(c.debug[pc])
+    result.stackTrace.reportInst = toReportLineInfo(instLoc())
+
+  finally:
+    # Clean up frames whenever leaving execution
+    c.sframes.setLen(0)
+
+proc unpackResult(res: sink ExecutionResult; config: ConfigRef, node: PNode; report = false): PNode =
+  ## Unpacks the execution result. If the result represents a failure, returns
+  ## a new `nkError` wrapping `node`. Otherwise the value/tree result is,
+  ## optionally filling in the node's `info` with that of `node`, if not
+  ## present already.
+  ##
+  ## `report` indicates whether instead of returning an nkError, the error
+  ## should be handled directly
+  if res.success:
+    result = res.result
+    if node != nil and result.info.line < 0:
+      result.info = node.info
+  else:
+    let errKind = res.report.kind
+
+    let stId = config.addReport(wrap(res.stackTrace))
+    let rId = config.addReport(wrap(res.report))
+
+    if not report:
```

Comment:
If you have both sides of the if, please prefer avoiding negation and simplify the boolean expression. It's easier to quickly scan/skim.